File: apps/pos_app/lib/data/models/human_authorization/canonical.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:pos_app/core/audit/v3/canonicalizer.dart' as audit;
import 'package:pos_app/core/audit/v3/scanner.dart' as audit;
import 'package:pos_app/core/audit/v3/sha256.dart' as audit;
import 'package:pos_app/core/audit/v3/types.dart' as audit;

import 'error_codes.dart';
import 'field_guards.dart';

/// Maps the shared audit-v3 scanner/canonicalizer codes onto the stable OHAC
/// codes so both runtimes reject the same byte-level problems identically.
///
/// Audit-v3 represents its codes as plain strings, so Dart cannot prove this
/// switch exhaustive at compile time. Every currently supported audit-v3 code
/// is therefore listed explicitly, and an unmapped future code throws instead
/// of silently degrading to OHAC_INVALID_JSON. The full mapping is pinned in
/// `canonical_test.dart`; extend both together when a new audit-v3 code ships.
String auditV3CodeToOhac(String auditCode) => switch (auditCode) {
      audit.auditV3InvalidUtf8 => OhacErrorCode.invalidUtf8,
      audit.auditV3InvalidJson => OhacErrorCode.invalidJson,
      audit.auditV3InvalidUnicode => OhacErrorCode.invalidUnicode,
      audit.auditV3DuplicateKey => OhacErrorCode.duplicateKey,
      audit.auditV3NumberForbidden => OhacErrorCode.numberForbidden,
      audit.auditV3LimitExceeded => OhacErrorCode.limitExceeded,
      audit.auditV3FrameInvalid => OhacErrorCode.invalidField,
      audit.auditV3FrameTooLarge => OhacErrorCode.limitExceeded,
      _ => throw StateError(
          'unmapped audit-v3 code "$auditCode": add an explicit OHAC mapping '
          'in canonical.dart and extend the pinning test instead of letting it '
          'silently become OHAC_INVALID_JSON',
        ),
    };

/// OHAC-C14N-1 adds two rules on top of the shared number-free canonicalizer:
/// `null` values are forbidden and no Unicode normalization is ever applied
/// (the underlying canonicalizer never normalizes).
bool _containsNull(audit.AuditV3Value value) => switch (value) {
      audit.AuditV3Null() => true,
      audit.AuditV3Array(:final values) => values.any(_containsNull),
      audit.AuditV3Object(:final entries) =>
        entries.any((entry) => _containsNull(entry.value)),
      _ => false,
    };

/// Canonicalizes raw UTF-8 JSON bytes with OHAC-C14N-1.
OhacResult<Uint8List> canonicalizeOhac(Uint8List rawUtf8) {
  final scanned = audit.scanNumberFreeJson(rawUtf8);
  if (scanned case final audit.AuditV3Failure<audit.AuditV3Value> failure) {
    return OhacFailure(OhacError(auditV3CodeToOhac(failure.error.code)));
  }
  final value = (scanned as audit.AuditV3Success<audit.AuditV3Value>).value;
  if (_containsNull(value)) {
    return const OhacFailure(OhacError(OhacErrorCode.nullForbidden));
  }
  final canonical = audit.canonicalizeNumberFreeJson(rawUtf8);
  if (canonical case final audit.AuditV3Failure<Uint8List> failure) {
    return OhacFailure(OhacError(auditV3CodeToOhac(failure.error.code)));
  }
  return OhacSuccess((canonical as audit.AuditV3Success<Uint8List>).value);
}

/// `sha256:` prefix plus 64 lowercase hex characters over canonical bytes.
String ohacDigest(Uint8List canonicalBytes) =>
    'sha256:${audit.sha256LowerHex(canonicalBytes)}';

/// Canonicalizes raw UTF-8 JSON bytes and returns their OHAC digest.
OhacResult<String> digestOfJson(Uint8List rawUtf8) {
  final canonical = canonicalizeOhac(rawUtf8);
  if (canonical case final OhacFailure<Uint8List> failure) {
    return OhacFailure<String>(failure.error);
  }
  return OhacSuccess<String>(
    ohacDigest((canonical as OhacSuccess<Uint8List>).value),
  );
}

/// OHAC-C14N-1 digests cover the whole object except `digest` itself, so the
/// transmitted digest can be verified against a body re-canonicalized from the
/// parsed value. Returns the body without the digest field on success.
OhacResult<Map<String, dynamic>> verifyBodyDigest(
  Map<String, dynamic> object,
) {
  final transmitted = object['digest'];
  if (!isDigest(transmitted)) {
    return OhacFailure(OhacError(OhacErrorCode.missingField, 'digest'));
  }
  final body = <String, dynamic>{
    for (final entry in object.entries)
      if (entry.key != 'digest') entry.key: entry.value,
  };
  final canonical = canonicalizeOhac(
    Uint8List.fromList(utf8.encode(jsonEncode(body))),
  );
  if (canonical case final OhacFailure<Uint8List> failure) {
    return OhacFailure<Map<String, dynamic>>(failure.error);
  }
  if (ohacDigest((canonical as OhacSuccess<Uint8List>).value) != transmitted) {
    return const OhacFailure(OhacError(OhacErrorCode.digestMismatch));
  }
  return OhacSuccess(body);
}

File: apps/admin_backend/src/modules/identity/human-authorization/contracts/canonical.ts
import { canonicalizeNumberFreeJson } from '../../../../core/audit/v3/canonicalizer';
import { scanNumberFreeJson } from '../../../../core/audit/v3/scanner';
import { sha256LowerHex } from '../../../../core/audit/v3/sha256';
import {
  AUDIT_V3_ERROR_CODE,
  type AuditV3Value,
  type AuditV3ErrorCode,
} from '../../../../core/audit/v3/types';
import {
  OHAC_ERROR_CODE,
  ohacFail,
  ohacOk,
  type OhacErrorCode,
  type OhacResult,
} from './error-codes';
import { isDigest } from './field-guards';

/**
 * OHAC-C14N-1: the project's existing number-free JCS subset, plus the two
 * rules v1 adds on top — no `null` values and no Unicode normalization.
 */
const AUDIT_TO_OHAC: Readonly<Record<AuditV3ErrorCode, OhacErrorCode>> = {
  [AUDIT_V3_ERROR_CODE.INVALID_UTF8]: OHAC_ERROR_CODE.INVALID_UTF8,
  [AUDIT_V3_ERROR_CODE.INVALID_JSON]: OHAC_ERROR_CODE.INVALID_JSON,
  [AUDIT_V3_ERROR_CODE.INVALID_UNICODE]: OHAC_ERROR_CODE.INVALID_UNICODE,
  [AUDIT_V3_ERROR_CODE.DUPLICATE_KEY]: OHAC_ERROR_CODE.DUPLICATE_KEY,
  [AUDIT_V3_ERROR_CODE.NUMBER_FORBIDDEN]: OHAC_ERROR_CODE.NUMBER_FORBIDDEN,
  [AUDIT_V3_ERROR_CODE.LIMIT_EXCEEDED]: OHAC_ERROR_CODE.LIMIT_EXCEEDED,
  [AUDIT_V3_ERROR_CODE.FRAME_INVALID]: OHAC_ERROR_CODE.INVALID_FIELD,
  [AUDIT_V3_ERROR_CODE.FRAME_TOO_LARGE]: OHAC_ERROR_CODE.LIMIT_EXCEEDED,
};

const containsNull = (value: AuditV3Value): boolean => {
  switch (value.kind) {
    case 'null':
      return true;
    case 'array':
      return value.values.some(containsNull);
    case 'object':
      return value.entries.some((entry) => containsNull(entry.value));
    default:
      return false;
  }
};

export const canonicalizeOhac = (rawUtf8: Buffer): OhacResult<Buffer> => {
  const scanned = scanNumberFreeJson(rawUtf8);
  if (scanned.ok === false) {
    return ohacFail(AUDIT_TO_OHAC[scanned.error.code]);
  }
  if (containsNull(scanned.value)) {
    return ohacFail(OHAC_ERROR_CODE.NULL_FORBIDDEN);
  }
  const canonical = canonicalizeNumberFreeJson(rawUtf8);
  if (canonical.ok === false) {
    return ohacFail(AUDIT_TO_OHAC[canonical.error.code]);
  }
  return ohacOk(canonical.value);
};

/** `sha256:` prefix plus 64 lowercase hex characters over canonical bytes. */
export const ohacDigest = (canonicalBytes: Buffer): string =>
  `sha256:${sha256LowerHex(canonicalBytes)}`;

export const digestOfJson = (rawUtf8: Buffer): OhacResult<string> => {
  const canonical = canonicalizeOhac(rawUtf8);
  if (canonical.ok === false) return canonical;
  return ohacOk(ohacDigest(canonical.value));
};

/**
 * OHAC-C14N-1 digests cover the whole object except `digest` itself, so the
 * transmitted digest can be verified against a body that is re-canonicalized
 * from the parsed value. Returns the body without the digest field on success.
 */
export const verifyBodyDigest = (
  object: Record<string, unknown>,
): OhacResult<Record<string, unknown>> => {
  const transmitted = object.digest;
  if (!isDigest(transmitted)) {
    return ohacFail(OHAC_ERROR_CODE.MISSING_FIELD, 'digest');
  }
  const body: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(object)) {
    if (key !== 'digest') body[key] = value;
  }
  const canonical = canonicalizeOhac(Buffer.from(JSON.stringify(body), 'utf8'));
  if (canonical.ok === false) return canonical;
  if (ohacDigest(canonical.value) !== transmitted) {
    return ohacFail(OHAC_ERROR_CODE.DIGEST_MISMATCH);
  }
  return ohacOk(body);
};
